src/common/survival_util.cc

```diff
@@ -0,0 +1,255 @@
+  
+  double pdf;
+  double w;
+  double hess;
+  pdf     = this->pdf(z);
+  w       = std::exp(z);
+  hess    = (w*w-3*w+1)*pdf;
+  return hess;
+}
+
+
+double AFTLoss::loss(double y_lower, double y_higher, double y_pred, double sigma) {
+  double pdf;
+  double cdf_u, cdf_l, z_u, z_l;
+  double cost;
+  if (y_lower == y_higher) {  // uncensored
+    z_l    = (y_lower-y_pred)/sigma;
+    pdf  = dist_->pdf(z_l);
+    cost = -std::log(pdf/(sigma*y_lower));
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_higher)) {  // right-censored
+      cdf_u  = 1;
+    } 
+    else{
+      z_u   = (y_higher-y_pred)/sigma;
+      cdf_u = dist_->cdf(z_u);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      cdf_l = 0;
+    } else {  // interval-censored
+      z_l   = (y_lower - y_pred)/sigma;
+      cdf_l = dist_->cdf(z_l);
+    } 
+    //else {
+    //  LOG(FATAL) << "AFTLoss: Could not determine event type: y_lower = " << y_lower
+    //             << ", y_higher = " << y_higher;
+    //}
+    cost = -std::log(cdf_u - cdf_l);
+  }
+  return cost;
+}
+
+double AFTLoss::gradient(double y_lower, double y_higher, double y_pred, double sigma) {
+  
+  double pdf_l;
+  double pdf_u;
+  double pdf;
+  double _grad;
+  double z;
+  double z_u;
+  double z_l;
+  double cdf_u;
+  double cdf_l;
+  double gradVar;
+  const double eps = 1e-12f;
+
+  if (y_lower == y_higher) {  // uncensored
+    z    = (y_lower-y_pred)/sigma;
+    pdf  = dist_->pdf(z);
+    _grad = dist_->grad_pdf(z);
+    gradVar = _grad/(sigma*pdf);
+  } else {  // censored; now check what type of censorship we have
+    if (std::isinf(y_higher)) {  // right-censored
+      pdf_u  = 0;
+      cdf_u  = 1;
+    } 
+    else{
+      z_l    = (y_lower-y_pred)/sigma;
+      pdf_l  = dist_->pdf(z_l);
+      cdf_l  = dist_->cdf(z_l);
+    }
+    if (std::isinf(y_lower)) {  // left-censored
+      pdf_l  = 0;
+      cdf_l  = 0;
+    } else {  // interval-censored
+      z_u    = (y_higher-y_pred)/sigma;
+      pdf_u  = dist_->pdf(z_u);
+      cdf_u  = dist_->cdf(z_u);
+    } 
+    //else {
+    //  LOG(FATAL) << "AFTLoss: Could not determine event type: y_lower = " << y_lower
+     //            << ", y_higher = " << y_higher;
+    //}
+    gradVar  = (pdf_u-pdf_l)/(sigma*std::max(cdf_u-cdf_l,eps));
+  }
+  
+  return gradVar;
+}
+
+double AFTLoss::hessian(double y_lower,double y_higher,double y_pred,double sigma){
+  double z; 
+  double z_u;
+  double z_l;
+  double pdf;
+  double pdf_u;
+  double pdf_l;
+  double cdf;
+  double cdf_u;
+  double cdf_l;
+  double grad_u;
+  double grad_l;
+  double gradVar;
+  double hessVar;
+  double hess_dist;
+  const double eps = 1e-12f;
+
+  if (y_lower == y_higher) {  // uncensored
+    z           = (y_lower-y_pred)/sigma;
+    pdf       = dist_->pdf(z);
+    gradVar   = dist_->cdf(z);
+    hess_dist = dist_->hess_pdf(z);
+    hessVar = -(pdf*hess_dist - std::pow(gradVar,2))/(std::pow(sigma,2)*std::pow(pdf,2));
```

Comment:
here as well you may consider using a call to gradient() for consistency with the surv pkg manual formula

Reply:
I have made the changes.